### PR TITLE
Change Vault Manager Params

### PR DESCRIPTION
**Pull request by samsiegart**
refs https://github.com/Agoric/dapp-econ-gov/issues/32

image

image

### CODE REVIEW

File: src/components/VaultParamChange.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+          />
+        );
+      default:
+        return <i className="text-gray-500">{type} not supported</i>;
+    }
+  }
+
+  function handleSubmit(event) {
+    event.preventDefault();
+    console.log({ event });
```
Comment (turadg) [inline src/components/VaultParamChange.tsx:99]:
```suggestion
    console.debug({ event });
```
or remove

File: src/components/VaultParamChange.tsx [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    const offer = walletUtils.makeVoteOnVaultManagerParams(
+      props.charterOfferId,
+      collateralBrand,
+      paramPatch,
+      minutesUntilClose,
+    );
+    void walletUtils.sendOffer(offer);
+  }
+
+  let content = (
```
Comment (turadg) [inline src/components/VaultParamChange.tsx:109]: consider a `const content` with a ternary

File: src/components/VaultParamChange.tsx [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+              onChange={e => setMinutesUntilClose(e.target.valueAsNumber)}
+            />
+          </label>
+          <div className="w-full flex flex-row justify-end mt-2">
+            <input
+              type="submit"
+              value="Propose Parameter Change"
+              className={clsx(
+                'btn-primary p-2 rounded mt-2',
+                canGovern ? 'cursor-pointer' : 'cursor-not-allowed',
```
Comment (turadg) [inline src/components/VaultParamChange.tsx:149]: TIL cursor-not-allowed

feels like overkill that we might not remember to do everywhere, but doesn't hurt

File: src/components/VaultsPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  charterInvitationSpec,
+  usePublishedDatum,
+  WalletContext,
+} from 'lib/wallet';
+import { useContext, useState } from 'react';
+import { FiChevronDown } from 'react-icons/fi';
+import VaultParamChange from './VaultParamChange';
+import Eligibility from './Eligibility';
+
+const ProposalTypes = {
```
Comment (turadg) [inline src/components/VaultsPanel.tsx:16]: thanks for showing where the code is going

File: src/components/inputs.tsx [hunk 5]
```diff
@@ -108,7 +117,6 @@ export const PercentageInput = ({
         onChange={handleInputChange}
         className="rounded bg-white bg-opacity-100 text-xl p-3 pr-10 leading-6 w-full border-gray-300 focus:border-purple-300 focus:ring-purple-300"
         min="0"
-        max="100"
```
Comment (turadg) [inline src/components/inputs.tsx:111]: consider an option to limit to ratios <=1.0

File: src/lib/rpc.ts [hunk 6]
```diff
@@ -10,7 +10,7 @@ import { makeFollower, makeLeader } from '@agoric/casting';
 
 export const networkConfigUrl = (agoricNetName: string) => {
   if (agoricNetName === 'local') {
-    return 'http://0.0.0.0:3000/wallet/network-config';
+    return 'https://wallet.agoric.app/wallet/network-config';
```
Comment (turadg) [inline src/lib/rpc.ts:13]: I don't follow why these RPC path changes are necessary. They really should be in separate commits than the UI changes.

File: src/components/inputs.tsx [hunk 7]
```diff
@@ -70,22 +70,31 @@ export const PercentageInput = ({
   ratio: Ratio;
   onChange: (newRatio: Ratio) => void;
 }) => {
-  assert(
-    ratio.denominator.value === 10_000n,
-    'only conventional denominator value supported',
-  );
+  const normalDenominator = 10_000n;
+  const scaleFactor = 10_000n / ratio.denominator.value;
```
[the remark this answers is not in the document]
Comment (samsiegart) [reply src/components/inputs.tsx]: Ah oh no:
```
tame-console.js:103 Error: "brand" {"boardId":"board0566","iface":"Alleged: IST brand"} must be a remotable, not "copyRecord"
  at PercentageInput
  ```
  Looks like the function expects a properly marshalled value, which we aren't doing.

Just changed it back, but also added a guard against divide-by-zero scenarios. Looks like the PSM always expected a denom of 10_000n, but our bootstrap config is giving some denoms of 100n, so this is just a workaround for that. Can we standardize on 10_000n to be safe?

File: src/components/VaultParamChange.tsx [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  // "Alleged: IbcATOM brand" -> "IbcATOM"
+  const collateralBrandLabel = collateralBrand?.iface?.split(' ')[1];
+
+  const [minutesUntilClose, setMinutesUntilClose] = useState(10);
+
+  const [paramPatch, setParamPatch] = useState({});
+
+  console.log('ProposeParamChange', { data, paramPatch });
+
+  const canMakeProposal = !!props.charterOfferId && collateralBrand;
```
Comment (turadg) [inline src/components/VaultParamChange.tsx:61]: if you want it boolean:
```suggestion
  const canMakeProposal = !!(props.charterOfferId && collateralBrand);
```

File: src/components/inputs.tsx [hunk 9]
```diff
@@ -66,26 +66,41 @@ export const AmountInput = ({
...
   onChange: (newRatio: Ratio) => void;
+  max?: string;
 }) => {
   assert(
-    ratio.denominator.value === 10_000n,
-    'only conventional denominator value supported',
+    ratio.denominator.value <= 10_000n,
+    'Cannot handle denominators > 10_000n',
   );
+  const normalDenominator = 10_000n;
```
Comment (turadg) [inline src/components/inputs.tsx:79]: should this be at module scope?

should definitely be before the `assert` to it can be used in it.
```js
assert(ratio.denominator.value <= normalDenominator, `Cannot handle denominators > ${normalDenominator`);
```

though… why can't it handle bigger denominators?

File: src/lib/rpc.ts (hunk 6 above)
Comment (turadg) [reply src/lib/rpc.ts:13]: the RPC paths seems like a different category of fix than UI tweaks to me. if we had downstream consumers of the we should convey the change through the commit history. not a blocker for now but something to keep in mind around communicating changes.

**Review by turadg — APPROVED** [review]

File: src/components/VaultParamChange.tsx (hunk 8 above)
Comment (samsiegart) [reply src/components/VaultParamChange.tsx]: Done thanks

File: src/components/inputs.tsx (hunk 9 above)
Comment (samsiegart) [reply src/components/inputs.tsx]: Sure just moved it to module scope. 

Since bigints cannot be decimals, the scaleFactor would become 0 if it was greater than 10,000. Also technically anything not divisible by 10 would be scaled inaccurately, but I'm assuming we won't have to worry about that. The implementation just gets a lot more complicated if we have to generalize to handle any denominator.

File: src/lib/rpc.ts (hunk 6 above)
Comment (samsiegart) [reply src/lib/rpc.ts:13]: Sure makes sense, I guess since it's not feature/bug. but rather development related.
